subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java

```diff
@@ -445,44 +491,39 @@ public void appendCachingSafeCopyAction(Action<? super FileCopyDetails> action)
         copyActions.add(action);
     }
 
+    @Override
     public CopySpec eachFile(Closure closure) {
         appendCopyAction(ConfigureUtil.configureUsing(closure));
         return this;
     }
 
+    @Override
     public Iterable<CopySpecInternal> getChildren() {
         return childSpecs;
     }
 
+    @Override
     public void walk(Action<? super CopySpecResolver> action) {
         buildRootResolver().walk(action);
     }
 
-    public boolean hasSource() {
-        if (!sourcePaths.isEmpty()) {
-            return true;
-        }
-        for (CopySpecInternal spec : childSpecs) {
-            if (spec.hasSource()) {
-                return true;
-            }
-        }
-        return false;
-    }
-
+    @Override
     public CopySpecResolver buildResolverRelativeToParent(CopySpecResolver parent) {
         return this.new DefaultCopySpecResolver(parent);
     }
 
+    @Override
     public CopySpecResolver buildRootResolver() {
         return this.new DefaultCopySpecResolver(null);
     }
 
+    @Override
     public String getFilteringCharset() {
         return buildRootResolver().getFilteringCharset();
     }
 
-    public void setFilteringCharset(String charset) {
+    @Override
+    public void setFilteringCharset(@Nullable String charset) {
```

Comment:
Can't we just remove the `@Nullable` here? We are throwing an exception when null is passed and the getter can't return null, either.

Reply:
That's a good point.